File: leetcode/17-letter-combinations-of-a-phone-number/main.go
package main

import "fmt"

/*
	1st approach: combinations, dfs in recursion
	1. for 13,a->d, a->e, a->f...so on
	2. quite similar to permutation problem

	Time	O(3^n*3^m)
	Space	O(3^n*3^m)
	n: number of 3 characters digits like 1->abc
	m: number of 4 characters digits like 9->wxyz
	0 ms, faster than 100.00%
*/
func letterCombinations(digits string) []string {
	mapping := make(map[byte][]string)
	mapping['2'] = []string{"a", "b", "c"}
	mapping['3'] = []string{"d", "e", "f"}
	mapping['4'] = []string{"g", "h", "i"}
	mapping['5'] = []string{"j", "k", "l"}
	mapping['6'] = []string{"m", "n", "o"}
	mapping['7'] = []string{"p", "q", "r", "s"}
	mapping['8'] = []string{"t", "u", "v"}
	mapping['9'] = []string{"w", "x", "y", "z"}

	res := []string{}

	var dfs func(prefix string, idx int)
	dfs = func(prefix string, idx int) {
		if idx == len(digits) {
			if len(prefix) > 0 {
				res = append(res, prefix)
			}
			return
		}
		digit := digits[idx]
		if v, x := mapping[digit]; x {
			for i := 0; i < len(v); i++ {
				dfs(prefix+v[i], idx+1)
			}
		} else {
			dfs(prefix, idx+1)
		}
	}
	dfs("", 0)
	return res
}

func main() {
	fmt.Println(letterCombinations(""))
	fmt.Println(letterCombinations("23"))
	fmt.Println(letterCombinations("13"))
	fmt.Println(letterCombinations("13#4"))
}

File: leetcode/17-letter-combinations-of-a-phone-number/main.py
"""
    1st approach: combinations, dfs in recursion
	1. for 13,a->d, a->e, a->f...so on
	2. quite similar to permutation problem

	Time	O(3^n -> 4^m)
	Space	O(3^n -> 4^m)
	n: number of 3 characters digits like 1->abc
	m: number of 4 characters digits like 9->wxyz
    20 ms, faster than 75.92%
"""


class Solution(object):

    def __init__(self):
        self.res = []

    def letterCombinations(self, digits):
        """
        :type digits: str
        :rtype: List[str]
        """
        ht = {}
        ht['2'] = 'abc'
        ht['3'] = 'def'
        ht['4'] = 'ghi'
        ht['5'] = 'jkl'
        ht['6'] = 'mno'
        ht['7'] = 'pqrs'
        ht['8'] = 'tuv'
        ht['9'] = 'wxyz'
        self.dfs("", digits, ht)
        return self.res

    def dfs(self, prefix, remain, ht):
        if len(remain) == 0 and len(prefix) > 0:
            self.res.append(prefix)
        elif len(remain) > 0:
            first = remain[0]
            if first not in ht:
                self.dfs(prefix, remain[1:], ht)
                return
            cands = ht[first]
            for cand in cands:
                self.dfs(prefix+cand, remain[1:], ht)


print(Solution().letterCombinations(""))
print(Solution().letterCombinations("23"))
print(Solution().letterCombinations("13"))
print(Solution().letterCombinations("13#4"))
print("-----")

"""
    1st approach: combinations, dfs in recursion
	1. for 13,a->d, a->e, a->f...so on
	2. quite similar to permutation problem

	Time	O(3^n -> 4^m)
	Space	O(3^n -> 4^m)
	n: number of 3 characters digits like 1->abc
	m: number of 4 characters digits like 9->wxyz
    24 ms, faster than 94.99%
"""


class Solution(object):

    def letterCombinations(self, digits):
        self.ht = {
            '2': 'abc',
            '3': 'def',
            '4': 'ghi',
            '5': 'jkl',
            '6': 'mno',
            '7': 'pqrs',
            '8': 'tuv',
            '9': 'wxyz',
        }
        self.res = []
        self.dfs(digits, '')
        return self.res

    def dfs(self, digits, chosen):
        if len(digits) == 0:
            if len(chosen) > 0:
                self.res.append(chosen)
            return
        num = digits[0]
        remain = digits[1:]
        # if num not in self.ht:
        #     self.dfs(digits[1:], chosen)
        #     return
        cands = self.ht[num]
        for cand in cands:
            self.dfs(remain, chosen + cand)


print(Solution().letterCombinations(""))
print(Solution().letterCombinations("23"))
print(Solution().letterCombinations("13"))
print(Solution().letterCombinations("13#4"))
